Copter: support zigzag auto feature

tatsuy opened the pull request:
In this PR, after setting the AB point in the ZigZag mode, turning on the switch automatically allows ZigZag and spraying.

I added these parameters and option.
ZIGZ_DIRECTION  0: forward, 1:right, 2: backward, 3: left
ZIGZ_LINE_NUM  Total line for ZigZag auto if 1 or more. -1: infinite, 0: Just moving to sideways
ZIGZ_SIDE_DIST (m) distance to move sideways
ZIGZ_WP_DELAY (s) delay for zigzag waypoint

Aux switch
RCx_OPTION 83 ZigZag Auto

And renamed from ZIGZAG_AUTO_PUMP to ZIGZ_SPRAYER.

The direction of movement is determined from the direction of Yaw. 
Therefore, do not rotate Yaw more than 90 degrees during ZigZag Auto. 

![image](https://user-images.githubusercontent.com/16643069/78744592-f32ea180-799c-11ea-9c5d-78757f87eb4d.png)

[How to use]
	1. set ZIGZ_\*, SPRAY_\*, RCx_OPTION 61(ZigZag SaveWP)/82(ZigZag Auto) and SERVOx_FUNCTION 22(SprayPump)/23(SpraySpinner) parameters
	2. Registered both A and B points
	3. Turn on ZigZag Auto using Aux Switch (RCx_OPTION = 83)

If you want to suspend, turn off ZigZag Auto switch.
It will resume after you set ZigZag Auto switch to High again.

This is the video of this feature in SITL.
https://youtu.be/MDKrd1XAAaI

There are screenshots in some test cases.
![image](https://user-images.githubusercontent.com/16643069/78744625-07729e80-799d-11ea-9d11-38e56fe648e6.png)
ZIGZ_DIRECTION 1 (Right)
ZIGZ_LINE_NUM 7
ZIGZ_SIDE_DIST 4 (4m)
ZIGZ_WP_DELAY 4 (4s)

![image](https://user-images.githubusercontent.com/16643069/78744706-32f58900-799d-11ea-96d7-529f1e1cdb92.png)
ZIGZ_DIRECTION 2 (Backward)
ZIGZ_LINE_NUM 4
ZIGZ_SIDE_DIST 7 (7m)
ZIGZ_WP_DELAY 1 (1s)

In the future work, I will create autotest and documentation.

Gone4Dirt commented:
Is this objectively different than what can be achieved with using auto missions?  I am struggling to see the advantage of using this vs setting an auto mission.

tatsuy commented:
Thank you for your comment.
I use both Auto mode and this mode at the spraying.

The difference is related to GPS accuracy.
The GPS position may have several meters error depending on the day.
When using the same mission in Auto mode, if the starting point shifts 3m to the left, the entire mission will also be offset about 3m to the left.
Due to this offset, there was a problem that it could not be sprayed to the intended place, such as spraying to different fields.
To solve this problem, I was offsetting missions via GCS during flight.

When using ZigZag mode, the starting point is set manually each time, so accuracy is better and I don't have to rewrite the mission during the flight.

Gone4Dirt commented:
That's interesting.  Have you tried or considered trying an rtk unit to improve positional accuracy? 
I can see that this is a useful addition for you.  My only concern is that it's a significant code addition with quite a few parameter additions for what is a bit of a niche use case.  Others that I know who do spraying have never mentioned such issues or requirements for such precision.
I appreciate that you have already put a lot of work into this already.  However, to suggest an alternative.  Have you considered using scripting to apply an offset to your entire mission based the difference from your planned start point to where you initiate auto mode? This would allow you to position the drone at the start point and use the pre existing mission tools, giving you the customisation that you desire, without having to add the code to the main code base and parameters.  Just a thought.
Would be interested to hear the opinions of others that do spraying work if this is would be helpful to them.

khancyr commented on ArduCopter/mode_zigzag.cpp:
```diff
@@ -120,39 +138,53 @@ void ModeZigZag::save_or_move_to_destination(uint8_t dest_num)
 
+void ModeZigZag::move_to_side()
+{
+    if (!dest_A.is_zero() && !dest_B.is_zero() && is_positive((dest_B - dest_A).length_squared())) {
```
is_positive((dest_B - dest_A).length_squared()) will always be true as squared number is always positive

tatsuy replied:
Thank you. I changed to use !is_zero.

khancyr commented:
I have never like zigzag mode as it mixes manual and auto mode. The use case is interesting but quite invasive even if restraint to the sole mode.
I wonder if this won't be better as scripting ? I thing that thing that should be done by planning on GCS shouldn't be put in the code, specially through the parameters.
I guess we will need to discuss this.

tatsuy commented:
@Gone4Dirt 
Thank you for your suggestion.
About RTK, I decided not to use RTK because it takes time to prepare and the cost.
At first I wanted to use Lua Script, but I gave it up because many copters are using CubeBlack.
>Others that I know who do spraying have never mentioned such issues or requirements for such precision.

This may be because the land in Japan is small and close to the neighboring fields.

tatsuy commented:
I wonder if most users using ZigZag mode would like to automate it.
It is convenient for users to complete their tasks with Transmitter only (don't need to create a mission in advance).

Gone4Dirt commented:
> At first I wanted to use Lua Script, but I gave it up because many copters are using CubeBlack.

Just for info for any future developments you may have, a CubeBlack has more than enough space to run quite sizeable scripts.  For example, I am confident it would be able to handle the mission offset routine that I previously suggested.

tatsuy commented:
Good information. I'll try it after a PR  to allow scripting to read and write missions has been merged.

rmackay9 commented on ArduCopter/mode_zigzag.cpp:
```diff
@@ -60,15 +60,34 @@ void ModeZigZag::run()
     pos_control->set_max_accel_z(g.pilot_accel_z);
 
+    // set the direction and the total number of lines
+        zigzag_direction = (Direction)constrain_int16(g2.zigzag_direction, 0, 3);
```
small white space issue here.

tatsuy replied:
fixed

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -967,14 +967,46 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Param: ZIGZ_SIDE_DIST
+    // @DisplayName: Sideways distance in ZigZag auto
+    // @Description: The distance to move sideways in ZigZag mode
+    // @Units: cm
```
let's change this to be in meters instead of cm

tatsuy replied:
ok, I changed it.

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -967,14 +967,46 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Units: ms
+    // @Range: 100 10000
+    // @User: Advanced
+    AP_GROUPINFO("ZIGZ_WP_DELAY", 39, ParametersG2, zigzag_wp_delay, ZIGZAG_WP_DELAY),
```
Could we change this delay to seconds instead of milliseconds?

tatsuy replied:
yes, I changed it.

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -967,14 +967,46 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Description: Total number of lines for ZigZag auto if 1 or more. -1: Infinity, 0: Disabled
+    // @Range: -1 32767
+    // @User: Advanced
+    AP_GROUPINFO("ZIGZ_LINE_NUM", 42, ParametersG2, zigzag_line_num, 0),
```
Would a maximum distance be better?  I'm not sure, just a question..

tatsuy replied:
It may be easier for users to input the size of the field..
But I select the total number of lines because I gave -1 and 0 special meaning.

tridge commented on ArduCopter/Parameters.cpp:
```diff
@@ -963,14 +963,46 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     // @User: Advanced
-    AP_GROUPINFO("ZIGZAG_AUTO_PUMP", 38, ParametersG2, zigzag_auto_pump_enabled, ZIGZAG_AUTO_PUMP_ENABLED),
-#endif
+    AP_GROUPINFO("ZIGZ_SPRAYER", 38, ParametersG2, zigzag_spray_enabled, ZIGZAG_SPRAY_ENABLED),
```
better to move these params to their own group

tatsuy replied:
I moved these to their own group

rmackay9 commented:
One suggestion to reduce the impact of the additional parameters is to move the param table to the flight mode class as is done in other modes like FlowHold (?) and SmartRTL.

tatsuy commented:
I've done two things to reduce the impact of the additional parameters:
1. moved ZIGZ_* params to flight mode class
2. added ZIGZ_AUTO_ENABLE parameter to be made invisible these parameters

```
> param show ZIGZ_*
> ZIGZ_AUTO_ENABLE 0.000000
param fetch
> Requested parameter list
Received 1238 parameters

> param set ZIGZ_AUTO_ENABLE 1
Received 1243 parameters
```

I confirmed that the existing ZigZag mode and the additional feature work in SITL.

rmackay9 commented:
I tested and rebased this PR here: https://github.com/rmackay9/rmackay9-ardupilot/tree/tatsuy-auto-zigzag

Also I added three commits to do the following:

- Moved the parameter indices to start from 1 instead of "0".  According to tridge we should be starting from "1" these days.  Hopefully this doesn't cause problems for any development users.
- Renamed "suspect" to "suspend_auto" to make it more clear what it refers to
- Minor formatting fixes

If you're happy with these changes can you pull them into this PR?

tatsuy commented:
@rmackay9 
Thank you for testing!
I pulled them and rebased.
I found another issue the problem that the default value is not set and I fixed it.
https://github.com/ArduPilot/ardupilot/blob/f45a09f6604792e18b7f1360bce113e89bd270cb/ArduCopter/mode_zigzag.cpp#L64

rmackay9 commented:
Great, merged, thanks!
